### PR TITLE
Build VLANs on fabric table

### PR DESCRIPTION
**Pull request by steverydz**
## Done
Built VLANs on fabric table

## QA
- Go to /MAAS/r/fabric/12
- Compare to the legacy version http://bolla.internal:5240/MAAS/l/fabric/12
- The content should be the same

## Issue
Fixes https://github.com/canonical-web-and-design/app-tribe/issues/645

### CODE REVIEW

File: ui/src/app/subnets/views/FabricDetails/FabricVLANs/FabricVLANs.test.tsx [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    </Provider>
+  );
+
+  expect(
+    screen.getByRole("heading", { name: "VLANs on this fabric" })
+  ).toBeInTheDocument();
+
+  expect(
+    screen.getByRole("columnheader", { name: "VLAN" })
+  ).toBeInTheDocument();
```
Comment (petermakowski) [inline ui/src/app/subnets/views/FabricDetails/FabricVLANs/FabricVLANs.test.tsx:34]: Would be great to test the actual content in these columns. You might need to add more data to the store first (subnets, vlans, etc.).

File: ui/src/app/subnets/views/FabricDetails/FabricVLANs/FabricVLANs.test.tsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+  expect(
+    screen.getByRole("columnheader", { name: "Subnet" })
+  ).toBeInTheDocument();
```
Comment (petermakowski) [inline ui/src/app/subnets/views/FabricDetails/FabricVLANs/FabricVLANs.test.tsx:38]: well done on using specific role queries 👍

**Comment by steverydz** [comment]
@petermakowski this is ready for re-review

File: ui/src/app/subnets/views/FabricDetails/FabricVLANs/FabricVLANs.test.tsx [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    screen.getByRole("columnheader", { name: "Available" })
+  ).toBeInTheDocument();
+
+  expect(
+    screen.getByRole("columnheader", { name: "Space" })
+  ).toBeInTheDocument();
+
+  expect(
+    screen.getByRole("gridcell", { name: /test-vlan/ })
+  ).toBeInTheDocument();
```
Comment (petermakowski) [inline ui/src/app/subnets/views/FabricDetails/FabricVLANs/FabricVLANs.test.tsx:85]: Although this doesn't test if the cell is in the correct column I think it's good enough for now 👍

File: ui/src/app/subnets/views/FabricDetails/FabricVLANs/FabricVLANs.tsx [hunk 4]
```diff
@@ -1,7 +1,149 @@
...
+  spaceId: Space["id"] | null
+): Space | undefined => {
+  return spaces.find((space) => space?.id === spaceId);
+};
+
+const getSubnetsInVLAN = (subnets: Subnet[], vlanId: VLAN["id"]): Subnet[] =>
+  subnets.filter((subnet) => subnet.vlan === vlanId);
+
+const getVLANsInFabric = (vlans: VLAN[], fabricId: Fabric["id"]): VLAN[] =>
+  vlans.filter((vlan) => vlan.fabric === fabricId);
```
Comment (petermakowski) [inline ui/src/app/subnets/views/FabricDetails/FabricVLANs/FabricVLANs.tsx:47]: You should be able to import most of the util functions you added to this file, they already exist, e.g. https://github.com/canonical-web-and-design/maas-ui/blob/9d880a47c3d8c5b980b10fca8862e3771141d982/ui/src/app/store/subnet/utils.ts#L51
Unless there's a reason why you duplicated them in this file?

File: ui/src/app/subnets/views/FabricDetails/FabricVLANs/FabricVLANs.tsx [hunk 5]
```diff
@@ -1,7 +1,149 @@
...
+  const spaces = useSelector(spaceSelectors.spaceState);
+
+  const data = getByFabric(
+    {
+      spaces: spaces.items,
+      subnets: subnets.items,
+      vlans: vlans.items,
+    },
+    fabric
+  );
```
Comment (petermakowski) [inline ui/src/app/subnets/views/FabricDetails/FabricVLANs/FabricVLANs.tsx:102]: You might want to wrap this in a `useMemo` / `useState`.
Without this the getByFabric function is be called (and data re-created) on each render. 

I'm doing a similar thing in the main subnets table: https://github.com/canonical-web-and-design/maas-ui/blob/9d880a47c3d8c5b980b10fca8862e3771141d982/ui/src/app/subnets/views/SubnetsList/SubnetsTable/hooks.ts#L42-L46

File: ui/src/app/subnets/views/FabricDetails/FabricVLANs/FabricVLANs.tsx [hunk 6]
```diff
@@ -1,7 +1,149 @@
...
+
+  const id = useId();
+
+  return (
+    <Strip shallow aria-labelledby={id}>
+      <h2 id={id} className="p-heading--4">
+        VLANs on this fabric
+      </h2>
```
Comment (petermakowski) [inline ui/src/app/subnets/views/FabricDetails/FabricVLANs/FabricVLANs.tsx:110]: You can use the new [TitledSection](https://github.com/canonical-web-and-design/maas-ui/blob/master/ui/src/app/base/components/TitledSection/TitledSection.tsx) component here instead.
```suggestion
    <TitledSection title="VLANs on this fabric">
```

**Review by petermakowski — COMMENTED** [review]
Looks good, just a few comments below.

**Comment by steverydz** [comment]
@petermakowski all comments addressed

**Review by Caleb-Ellis — APPROVED** [review]
Thanks @steverydz! This looks and works great. I think some of these util functions could be turned into one or more selectors, but otherwise LGTM!

**Pull request merged**